### PR TITLE
Move BIR serialization to project api

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
Previously we have serialized the bir into binary form during compilation (birgen) so there is no control from project api side. Then we store the serialized byte content in the BPackageSymbol. We already store the birBytes in ModuleContext as well. In this PR I have moved serialization to project api and make the getBirBytes function to have on-demand serialization. 

I also moved the bir writer to project api as well. 

Resolves https://github.com/ballerina-platform/ballerina-lang/issues/28289

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java [hunk 1]
```diff
@@ -133,6 +137,13 @@ ModuleName moduleName() {
         return moduleDescriptor.name();
     }
 
+    byte[] getBirBytes() {
+        if (birBytes == null) {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:141]: function to get bit bytes.

File: tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/bala/BalaCreator.java [hunk 2]
```diff
@@ -179,7 +179,7 @@ public static CompileResult compileWithTestsAndWrite(Path sourceRootPath, String
 
         BLangPackage bLangPackage = (BLangPackage) compileResult.getAST();
         Compiler compiler = Compiler.getInstance(context);
-        compiler.write(bLangPackage, targetPath);
+        //compiler.write(bLangPackage, targetPath);
```
Comment (warunalakshitha) [inline tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/bala/BalaCreator.java:182]: This class already deprecated but still, some functions used. We need to use use {@link org.ballerinalang.test.BCompileUtil} instead.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java [hunk 3]
```diff
@@ -133,6 +137,13 @@ ModuleName moduleName() {
...
+            birBytes = new BIRBinaryWriter(this.bLangPackage.symbol.bir).serialize();
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:142]: this.bLangPackage can be `null`, let's use `getBLangPackageOrThrow` so we get a better error message.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java [hunk 4]
```diff
@@ -87,6 +87,10 @@ public ModuleDescriptor descriptor() {
         return new DocumentIterable(this.testSrcDocs.values());
     }
 
+    public byte[] birBytes() {
+        return this.moduleContext.getBirBytes();
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java:91]: I don't see any other compiler outputs (such as getBLangPackageOrThrow) exposed from this class, Isn't it enough to expose this form the `ModuleContext`

Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java]: I have removed exposing bir from Module. This was only used for test bir. Added api in nullbackend (which is use for test bir purpose) to get default module bir.

**Comment by warunalakshitha** [comment]
There are conflicts. Will fix and send a new PR.

**Pull request closed without merging**